# Python Logic Statements

## logic operators

*  greater than >
*  less than <
*  equal to == 
    * a single "=" sign means to assign
*  not equal to !=
*  greater than or equal >=
*  less than or equal to <=

In [1]:
a = 42  # 2800
b = 47  # 572
c = 27  # 1750 
d = 42  # 2801

In [2]:
a > b

False

In [3]:
response = a > b

In [4]:
response

False

In [5]:
b > c

True

In [6]:
a > d

False

In [7]:
a >= d

True

In [8]:
a < d

False

In [9]:
a <= d

True

In [10]:
a == d

True

In [12]:
(a != d)

False

## dual logic statements  - and, or

In [13]:
(a/b > d) & (b/c < a)

False

In [14]:
(a/b > d) and (b/c < a)

False

In [15]:
(a/b < d) & (b/c < a)

True

In [16]:
(a/b < d) and (b/c < a)

True

In [17]:
(a/b > d) | (b/c < a)

True

In [18]:
(a/b > d) or (b/c < a)

True

## Conditionals if, else, elif

We will now look at conditional statements
There are 3 types of conditionals
1. if statement
2. else statement
3. elif statement

## if

In [20]:
if True:
  print('true')

true


In [21]:
if 1:
  print('true')

true


In [22]:
if False:
  print('false')

In [23]:
if 0:
  print('false')

In [25]:
if a > b:
  print("{} is great than {}".format(a, b))

In [27]:
if a < b:
  print("{} is less than {}".format(a, b))

42 is less than 47


In [29]:
if a < b:
  print(f"{a} is less than {b}")

42 is less than 47


In [30]:
if (a < b) & (a < d):
  print(f'{a} is less than {b} and {c}')

In [31]:
if (a < b) & (a <= d):
  print(f'{a} is less than {b} and {c}')

42 is less than 47 and 27


## else

In [32]:
if a > b:
  print(f"{a} is greater than {b}")

In [33]:
if a > b:
  print(f"{a} is greater than {b}")
else:
  print(f"{a} is less than {b}")

42 is less than 47


## elif

In [35]:
if a > d:
  print(f'{a} is greater than {d}')
elif a < d:
  print(f'{a} is less than {d}')
else:
  print(f'{a} is equal to {d}')

42 is equal to 42


## one line if statement

In [39]:
x = 1

In [40]:
if x == 1: print('hi')

hi


## one line else if

In [43]:
x = 0

In [44]:
print('hi') if x == 1 else print('not hi')

not hi


# working with the data

## import the data

In [45]:
import pandas as pd

In [ ]:
# from google.colab import drive
# drive.mount('/content/gdrive')

In [ ]:
# df = pd.read_csv('/content/gdrive/MyDrive/01. Projects/02. DataSimple Classes/03. Lesson Plans/07. Data Analysis Bootcamp/02. Logic Statements, Filtering, Selecting Data in Pandas/dino.csv')

In [46]:
df = pd.read_csv('https://raw.githubusercontent.com/brandynewanek/data/main/dino.csv')

In [47]:
df.head()

height  teeth size  tail length  dino
0   3.244112   18.865971    13.713091     2
1  15.391455    2.115573    20.942076     3
2  14.625879    4.168479    22.010666     3
3   5.574288   21.108127    14.003040     2
4  15.675385    6.361126     7.721574     1

### How many types of dinos are there? value_counts()

In [48]:
df['dino'].value_counts()

2    250
3    250
1    250
0    250
Name: dino, dtype: int64

250 observations of each type of dino

### How tall is dino type 1?

In [49]:
df['dino'] == 1

0      False
1      False
2      False
3      False
4       True
       ...  
995    False
996     True
997     True
998     True
999    False
Name: dino, Length: 1000, dtype: bool

In [50]:
df.shape

(1000, 4)

In [51]:
df[df['dino'] == 1].shape

(250, 4)

In [52]:
df_dino1 = df[df['dino'] == 1].copy()

In [53]:
df_dino1.head()

height  teeth size  tail length  dino
4   15.675385    6.361126     7.721574     1
8   15.767728    5.602762     5.343678     1
12  12.593850    5.389443     5.086763     1
13  14.429057    8.285375     4.476372     1
15  13.731934    6.472428     3.868351     1

In [54]:
df_dino1['height'].mean()

14.054850710391307

### What is the count of each type of dinosaur with teeth size between 16 and 22?

In [55]:
df[(df['teeth size'] > 16) & (df['teeth size'] < 22)]

height  teeth size  tail length  dino
0     3.244112   18.865971    13.713091     2
3     5.574288   21.108127    14.003040     2
6     9.590454   21.510810    15.089215     0
7     7.765885   21.451999    15.627048     0
11    1.950500   21.054287    14.036573     2
..         ...         ...          ...   ...
992   3.128442   20.115406    13.893763     2
993   1.392596   19.628810    13.613018     2
994  11.028839   21.978460    18.204522     0
995   9.332794   21.587405    15.627774     0
999   3.013892   19.870275    15.474768     2

[357 rows x 4 columns]

In [56]:
df[(df['teeth size'] > 16) & (df['teeth size'] < 22)]['dino'].value_counts()

2    241
0    116
Name: dino, dtype: int64

pandas does not work with the writting and logic operator only the &

In [59]:
df[(df['teeth size'] > 16) and (df['teeth size'] < 22)]

ValueError: ignored

### Which dinos have smaller than 5 tail length or greater than 20?

In [60]:
(df['tail length'] < 5) | (df['tail length'] > 20)

0      False
1       True
2       True
3      False
4      False
       ...  
995    False
996    False
997    False
998     True
999    False
Name: tail length, Length: 1000, dtype: bool

In [61]:
df[(df['tail length'] < 5) | (df['tail length'] > 20)]

height  teeth size  tail length  dino
1    15.391455    2.115573    20.942076     3
2    14.625879    4.168479    22.010666     3
13   14.429057    8.285375     4.476372     1
15   13.731934    6.472428     3.868351     1
18   16.502041    3.802386    20.071725     3
..         ...         ...          ...   ...
969  18.525324    2.625704    20.017152     3
975  13.833580    7.228556     4.079984     1
977  15.774352    4.107228    22.247299     3
981  18.032417    3.801304    20.529904     3
998  14.929872    5.334383     3.788657     1

[333 rows x 4 columns]

In [63]:
df[(df['tail length'] < 5) | (df['tail length'] > 20)]['dino'].value_counts()

3    228
1    105
Name: dino, dtype: int64

also pandas does not work with the written or only the |

In [64]:
df[(df['tail length'] < 5) or (df['tail length'] > 20)]

ValueError: ignored

In [65]:
(df['tail length'] < 5) | (df['tail length'] > 20)

0      False
1       True
2       True
3      False
4      False
       ...  
995    False
996    False
997    False
998     True
999    False
Name: tail length, Length: 1000, dtype: bool